io/src/main/scala/fs2/io/tcp/Socket.scala

```diff
@@ -0,0 +1,305 @@
+package fs2.io.tcp
+
+
+import java.net.{StandardSocketOptions, InetSocketAddress, SocketAddress}
+import java.nio.ByteBuffer
+import java.nio.channels.spi.AsynchronousChannelProvider
+import java.nio.channels.{AsynchronousCloseException, AsynchronousServerSocketChannel, CompletionHandler, AsynchronousSocketChannel, AsynchronousChannelGroup}
+import java.util.concurrent.TimeUnit
+
+import fs2.Chunk.Bytes
+import fs2._
+import fs2.Stream._
+
+import scala.concurrent.duration._
+
+
+trait Socket[F[_]] {
+
+  /**
+    * Read up to `maxBytes` from the peer.
+    *
+    * Evaluates to None, if there are no more bytes to be read in future, due stream reached End-Of-Stream state
+    * before returning even single byte. Otherwise returns Some(bytes) with bytes that were ready to be read.
+    *
+    * If `timeout` is specified, then resulting `F` will evaluate to failure with [[java.nio.channels.InterruptedByTimeoutException]]
+    * if read was not satisfied in given timeout. Read is satisfied, when at least single Byte was received
+    * before `timeout` expires
+    *
+    * This may return None, as well when end of stream has been reached before timeout expired and no data
+    * has been received.
+    *
+    */
+  def read(
+    maxBytes: Int
+    , timeout: Option[FiniteDuration] = None
+  ): F[Option[Bytes]]
+
+  /**
+    * Reads stream of bytes from this socket with `read` semantics. Terminates when eof is received.
+    * On timeout, this fails with [[java.nio.channels.InterruptedByTimeoutException]].
+    */
+  def reads(
+    maxBytes:Int
+    , timeout: Option[FiniteDuration] = None
+  ):Stream[F,Bytes]
+
+
+
+  /**
+    * Read exactly `numBytes` from the peer in a single chunk.
+    * If `timeout` is provided and no data arrives within the specified duration, then this results in
+    * failure with [[java.nio.channels.InterruptedByTimeoutException]]
+    *
+    * When returned size of bytes is < `numBytes` that indicates end-of-stream has been reached.
```

Comment:
Its perhaps my poor english, but I understand the comment that way. Read always `N` bytes (wait for them to be available). Only receive < N on eof. Fail on timeout if provided.  

`read` in contrary takes what is availbale (up to the maximum) possibly None if nothing is available at all. I was in fact thinking to get rid of Option and just user empty chunk of bytes, but I think we sort of agreed on the Api. However if you have idea to changing is no brainer....